/sshnpd/srv/src/srv.c:273:13: error: incompatible pointer to integer conversion initializing 'pthread_t' (aka 'unsigned long') with an expression of type 'void *' [-Wint-conversion]

### Describe the bug

When building C sshnpd for linux/amd64 (glibc) as [action](https://github.com/atsign-foundation/noports/actions/runs/11797328778/job/32861125353#step:5:350) for release c0.2.4

```log
#9 15.95 [ 38%] Building C object _deps/srv-lib-build/CMakeFiles/srv-lib.dir/src/srv.c.o
#9 16.01 /sshnpd/srv/src/srv.c:273:13: error: incompatible pointer to integer conversion initializing 'pthread_t' (aka 'unsigned long') with an expression of type 'void *' [-Wint-conversion]
#9 16.01   273 |   pthread_t tid = NULL;
#9 16.01       |             ^     ~~~~
#9 16.01 1 error generated.
#9 16.01 make[2]: *** [_deps/srv-lib-build/CMakeFiles/srv-lib.dir/build.make:104: _deps/srv-lib-build/CMakeFiles/srv-lib.dir/src/srv.c.o] Error 1
#9 16.01 make[1]: *** [CMakeFiles/Makefile2:3375: _deps/srv-lib-build/CMakeFiles/srv-lib.dir/all] Error 2
#9 16.01 make: *** [Makefile:136: all] Error 2
```

### Steps to reproduce

1. First I created a new release c0.2.4 to incorporate #1521 and #1522
2. Then the C_Release_Automation workflow was triggered and run
3. And then the [docker_build (linux/amd64, sshnpd-linux-x64)](https://github.com/atsign-foundation/noports/actions/runs/11797328778/job/32861125353#logs) job failed emitting the log snippet pasted above

### Expected behavior

Clean build with no warnings

### Additional context

[musl_build (linux/amd64, sshnpd-linux-x64-musl)](https://github.com/atsign-foundation/noports/actions/runs/11797328778/job/32861126126#logs) succeeded, so this likely relates to gcc and glibc

--- a/packages/c/srv/src/srv.c
+++ b/packages/c/srv/src/srv.c
@@ -298,6 +298,7 @@ int socket_to_socket(const srv_params_t *params, const char *auth_string, chunke
 
   res = pthread_create(&threads[1], NULL, srv_side_handle, &sides[1]);
   if (res != 0) {
+    atlogger_log(TAG, ERROR, "Failed to create thread: 1\n");
     cancel_first = true;
     exit_res = res;
     goto cancel;

--- a/packages/c/srv/src/srv.c
+++ b/packages/c/srv/src/srv.c
@@ -269,8 +269,8 @@ int socket_to_socket(const srv_params_t *params, const char *auth_string, chunke
 
   int fds[2], tidx;
   int exit_res = 0;
-  pthread_t threads[2];
-  pthread_t tid = NULL;
+  pthread_t threads[2], tid;
+  bool cancel_first = false;
   pipe(fds);
 
   srv_link_sides(&sides[0], &sides[1], fds);
@@ -298,7 +298,7 @@ int socket_to_socket(const srv_params_t *params, const char *auth_string, chunke
 
   res = pthread_create(&threads[1], NULL, srv_side_handle, &sides[1]);
   if (res != 0) {
-    atlogger_log(TAG, ERROR, "Failed to create thread: 1\n");
+    cancel_first = true;
     exit_res = res;
     goto cancel;
   }
@@ -316,10 +316,13 @@ int socket_to_socket(const srv_params_t *params, const char *auth_string, chunke
   read(fds[0], &tid, sizeof(pthread_t));
 
   atlogger_log(TAG, DEBUG, "Joining exited thread\n");
+
+  // When a thread exits, join it.
   res = pthread_join(tid, (void *)&retval);
 
 cancel:
-  if (pthread_equal(threads[0], tid) > 0) {
+  // Then figure out which thread didn't close
+  if (!cancel_first && pthread_equal(threads[0], tid) > 0) {
     // If threads[0] exited normally then we will cancel threads[1]
     // In all other cases, cancel threads[0] (could be because threads[1] exited or errored)
     tidx = 1;
@@ -327,6 +330,7 @@ cancel:
     tidx = 0;
   }
 
+  // Then cancel the other thread
   atlogger_log(TAG, DEBUG, "Cancelling remaining open thread: %d\n", tidx);
   if (pthread_cancel(threads[tidx]) != 0) {
     atlogger_log(TAG, WARN, "Failed to cancel thread: %d\n", tidx);

